extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -0,0 +1,232 @@
+
+    const props = this.config_['optout'].split('.');
+    let obj = this.getWin();
+    for (let i = 0; i < props.length; i++) {
+      if (!obj) {
+        return false;
+      }
+      obj = obj[props[i]];
+    }
+    return obj();
+
+  }
+  /**
+   * Goes through all the requests in predefined vendor config and tag's config
+   * and creates a map of request name to request template. These requests can
+   * then be used while sending a request to a server.
+   *
+   * @private
+   */
+  generateRequests_() {
+    const requests = {};
+    if (!this.config_ || !this.config_['requests']) {
+      log.error(this.getName_(), 'No request strings defined. Analytics data ' +
+          'will not be sent from this page.');
+      return;
+    }
+
+    for (const k in this.config_['requests']) {
+      if (this.config_['requests'].hasOwnProperty(k)) {
+        requests[k] = this.config_['requests'][k];
+      }
+    }
+    this.requests_ = requests;
+
+    // Now replace any request templates with their values.
+    const all = Object.keys(this.requests_).join('|');
+    const requestExpr = new RegExp('\{(' + all + ')\}', 'g');
+    for (const k in this.requests_) {
+      this.requests_[k] = this.expandRequest_(this.requests_[k], requestExpr,
+          this.requests_, 5);
+    }
+  }
+
+  /**
+   * Replaces template variables corresponding to a request with the value of
+   * that request. If no request is found, an empty string is inserted instead.
+   *
+   * @param {string} formatString The request string to be expanded
+   * @param {RegExp} expression The regular expression used to detect request
+   *          templates
+   * @param {Object.<string, string> requests Map of requests to lookup template
+   *          values from
+   * @param {number} depth Depth of recursion for nested templates
+   * @return {string} the expanded request string
+   * @private
+   */
+  expandRequest_(formatString, expression, requests, depth) {
+    if (depth <= 0) {
+      return '';
+    }
+    depth--;
+    return formatString.replace(expression, (match, name) => {
+      return this.expandRequest_(requests[name], expression, requests, depth) ||
+          '';
+    });
+  }
+
+  /**
+   * Callback for events that are registered by the config's triggers. This
+   * method generates the request and sends the request out.
+   *
+   * @param {!JSONObject} trigger JSON config block that resulted in this event.
+   * @param {!Object} event Object with details about the event.
+   * @private
+   */
+  handleEvent_(trigger, event) {
+    const baseConfig = AnalyticsConfig[this.type_] || {};
+    const host = this.config_['host'] || baseConfig['host'];
+    let request = this.requests_[trigger['request']];
+    if (!host || !request) {
+      return;
+    }
+
+    // TODO(btownsend, #871): Add support for variables from inline config.
+    request = urlReplacementsFor(this.getWin()).expand(request);
+
+    // TODO(btownsend, #871): Add support for sendBeacon.
```

Comment:
Please create a separate issue for this. E.g. "Use sendBeacon for amp-analytics".